slotsmanager.c

```diff
@@ -0,0 +1,427 @@
+            int add = 0.2 * newMaxSlots[ptype];
+            if (add < 128) add = 128;
+            newMaxSlots[ptype] += add;
+            good = 0;
+        }
+    }
+    /* no need to grow, already have enough */
+    if (good) {
+        return;
+    }
+    /* FIXME: do a global max; because all variables in All.* are synced between ranks. */
+
+    size_t total_bytes = 0;
+    size_t offsets[6];
+    size_t bytes[6] = {0};
+
+    for(ptype = 0; ptype < 6; ptype++) {
+        offsets[ptype] = total_bytes;
+        bytes[ptype] = SlotsManager->info[ptype].elsize * newMaxSlots[ptype];
+        total_bytes += bytes[ptype];
+    }
+    char * newSlotsBase = myrealloc(SlotsManager->Base, total_bytes);
+
+    message(1, "Allocated %g MB for %d sph, %d stars and %d BHs.\n", total_bytes / (1024.0 * 1024.0),
+            newMaxSlots[0], newMaxSlots[4], newMaxSlots[5]);
+
+    /* move the last block first since we are only increasing sizes, moving items forward */
+    for(ptype = 5; ptype >= 0; ptype--) {
+        memmove(newSlotsBase + offsets[ptype],
+            SlotsManager->info[ptype].ptr,
+            SlotsManager->info[ptype].elsize * SlotsManager->info[ptype].size);
+    }
+
+    SlotsManager->Base = newSlotsBase;
+
+    for(ptype = 0; ptype < 6; ptype++) {
+        SlotsManager->info[ptype].ptr = newSlotsBase + offsets[ptype];
+        SlotsManager->info[ptype].maxsize = newMaxSlots[ptype];
+    }
+    GDB_SphP = (struct sph_particle_data *) SlotsManager->info[0].ptr;
+    GDB_StarP = (struct star_particle_data *) SlotsManager->info[4].ptr;
+    GDB_BhP = (struct bh_particle_data *) SlotsManager->info[5].ptr;
+}
+
+void slots_init()
+{
+    int ptype;
+    memset(SlotsManager, 0, sizeof(SlotsManager[0]));
+
+    SlotsManager->info[0].elsize = sizeof(struct sph_particle_data);
+    SlotsManager->info[0].enabled = 1;
+    SlotsManager->info[4].elsize = sizeof(struct star_particle_data);
+    SlotsManager->info[4].enabled = 1;
+    SlotsManager->info[5].elsize = sizeof(struct bh_particle_data);
+    SlotsManager->info[5].enabled = 1;
+
+    MPI_Type_contiguous(sizeof(struct particle_data), MPI_BYTE, &MPI_TYPE_PARTICLE);
+    MPI_Type_commit(&MPI_TYPE_PARTICLE);
+
+    for(ptype = 0; ptype < 6; ptype++) {
+        if(SlotsManager->info[ptype].enabled) {
+            MPI_Type_contiguous(SlotsManager->info[ptype].elsize, MPI_BYTE, &MPI_TYPE_SLOT[ptype]);
+            MPI_Type_commit(&MPI_TYPE_SLOT[ptype]);
+        }
+    }
+}
+
+void
+slots_check_id_consistency()
+{
+    int used[6] = {0};
+    int i;
+
+#pragma omp parallel for
+    for(i = 0; i < NumPart; i++) {
+        if(!SlotsManager->info[P[i].Type].enabled) continue;
+
+        if(P[i].PI >= SlotsManager->info[P[i].Type].size) {
+            endrun(1, "slot PI consistency failed2\n");
+        }
+        if(BASESLOT(i)->ID != P[i].ID) {
+            endrun(1, "slot id consistency failed2\n");
+        }
+#pragma omp atomic
+        used[P[i].Type] ++;
```

Comment:
This variable is unused. Did you mean to use it for something?

Reply:
I thought it was to be printed in the DEBUG mode...